numpy/lib/tests/test_index_tricks.py

```diff
@@ -249,6 +249,26 @@ def test_mgrid_size_none_handling(self, start, stop, step, expected):
         assert_equal(grid.size, expected[0])
         assert_equal(grid_small.size, expected[1])
 
+    def test_accepts_npfloating(self):
+        # regression test for #16466
+        grid64 = mgrid[0.1:0.33:0.1, ]
+        grid32 = mgrid[np.float32(0.1):np.float32(0.33):np.float32(0.1), ]
+        assert_(grid32.dtype == np.float64)
+        assert_array_almost_equal(grid64, grid32)
+
+        # different code path for single slice
+        grid64 = mgrid[0.1:0.33:0.1]
+        grid32 = mgrid[np.float32(0.1):np.float32(0.33):np.float32(0.1)]
+        assert_(grid32.dtype == np.float64)
+        assert_array_almost_equal(grid64, grid32)
+
+    def test_accepts_npcomplexfloating(self):
+        grid = mgrid[0.1:0.3:np.complex64(3j), ]
+        assert_array_almost_equal(grid, np.array([[0.1, 0.2, 0.3]]))
+
+        # different code path for single slice
+        grid = mgrid[0.1:0.3:np.complex64(3j)]
+        assert_array_almost_equal(grid, np.array([0.1, 0.2, 0.3]))
```

Comment:
```suggestion
        assert_array_almost_equal(
            mgrid[0.1:0.3:3j, ], mgrid[0.1:0.3:np.complex64(3j), ]
        )

        # different code path for single slice
        assert_array_almost_equal(
            mgrid[0.1:0.3:3j], mgrid[0.1:0.3:np.complex64(3j)]
        )
```
I think these tests should be reformulated so that they match the `float` versions above, i.e. directly compare the results of mgrid with complex and complex64 steps, rather than comparing to the *output* of mgrid.